[TEST-FAILURE] SplitBrainHandlerTest failures

com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge
com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave
```
java.lang.AssertionError: expected:<Address[127.0.0.1]:5703> but was:<Address[127.0.0.1]:5702>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge(SplitBrainHandlerTest.java:710)
```
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.maintenance-nightly/474/#showFailuresLink

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java
@@ -604,15 +604,19 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
             return;
         }
 
-        if (!node.isMaster()) {
-            sendMasterAnswer(joinRequest.getAddress());
+        Address target = joinRequest.getAddress();
+        boolean isRequestFromCurrentMaster = target.equals(node.getMasterAddress());
+        // If the join request from current master, don't send a master answer
+        // because master can somehow dropped its connection and wants to join back.
+        if (!node.isMaster() && !isRequestFromCurrentMaster) {
+            sendMasterAnswer(target);
             return;
         }
 
         if (joinInProgress) {
             if (logger.isFinestEnabled()) {
                 logger.finest("Join is in-progress. Cannot handle join request from "
-                        + joinRequest.getAddress() + " at the moment.");
+                        + target + " at the moment.");
             }
             return;
         }
@@ -625,12 +629,12 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
 
             long now = Clock.currentTimeMillis();
             if (logger.isFinestEnabled()) {
-                String msg = "Handling join from " + joinRequest.getAddress() + ", inProgress: " + joinInProgress
+                String msg = "Handling join from " + target + ", inProgress: " + joinInProgress
                         + (timeToStartJoin > 0 ? ", timeToStart: " + (timeToStartJoin - now) : "");
                 logger.finest(msg);
             }
 
-            MemberInfo memberInfo = new MemberInfo(joinRequest.getAddress(), joinRequest.getUuid(),
+            MemberInfo memberInfo = new MemberInfo(target, joinRequest.getUuid(),
                     joinRequest.getAttributes());
 
             if (!setJoins.contains(memberInfo)) {
@@ -638,7 +642,7 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
                     checkSecureLogin(joinRequest, memberInfo);
                 } catch (Exception e) {
                     ILogger securityLogger = node.loggingService.getLogger("com.hazelcast.security");
-                    sendAuthenticationFailure(joinRequest.getAddress());
+                    sendAuthenticationFailure(target);
                     securityLogger.severe(e);
                     return;
                 }
@@ -649,7 +653,7 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
                     node.getNodeExtension().beforeJoin();
                 } catch (Exception e) {
                     logger.warning(e.getMessage());
-                    sendBeforeJoinCheckFailure(joinRequest.getAddress(), e.getMessage());
+                    sendBeforeJoinCheckFailure(target, e.getMessage());
                     return;
                 }
             }
@@ -659,7 +663,7 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
             }
 
             if (setJoins.add(memberInfo)) {
-                sendMasterAnswer(joinRequest.getAddress());
+                sendMasterAnswer(target);
                 if (now - firstJoinRequest < maxWaitMillisBeforeJoin) {
                     timeToStartJoin = now + waitMillisBeforeJoin;
                 }

--- a/hazelcast/src/test/java/com/hazelcast/cluster/SplitBrainHandlerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/SplitBrainHandlerTest.java
@@ -41,12 +41,6 @@ import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.NightlyTest;
 import com.hazelcast.util.Clock;
-import org.junit.After;
-import org.junit.Before;
-import org.junit.Test;
-import org.junit.experimental.categories.Category;
-import org.junit.runner.RunWith;
-
 import java.io.IOException;
 import java.nio.channels.ServerSocketChannel;
 import java.util.Arrays;
@@ -60,9 +54,13 @@ import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
 
 import static com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance;
-import static com.hazelcast.test.HazelcastTestSupport.closeConnectionBetween;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;

